common/src/main/java/com/wynntils/core/text/StyledText.java

```diff
@@ -0,0 +1,487 @@
+
+        PartStyle previousStyle = null;
+        for (StyledTextPart part : parts) {
+            builder.append(part.getString(previousStyle, type));
+            previousStyle = part.getPartStyle();
+        }
+
+        return builder.toString();
+    }
+
+    public MutableComponent getComponent() {
+        if (parts.isEmpty()) {
+            return Component.empty();
+        }
+
+        MutableComponent component = parts.get(0).getComponent();
+
+        for (int i = 1; i < parts.size(); i++) {
+            component.append(parts.get(i).getComponent());
+        }
+
+        return component;
+    }
+
+    public static StyledText join(String separator, StyledText... texts) {
+        List<StyledTextPart> parts = new LinkedList<>();
+        List<ClickEvent> clickEvents = new LinkedList<>();
+        List<HoverEvent> hoverEvents = new LinkedList<>();
+
+        final int length = texts.length;
+        for (int i = 0; i < length; i++) {
+            StyledText text = texts[i];
+            parts.addAll(text.parts);
+
+            if (i != length - 1) {
+                // Possible micro-optimization: If the separator does not have formatting, use the normal constructor
+                parts.add(StyledTextPart.fromComponentWithPossibleFormattingCodes(separator, Style.EMPTY, null, null));
+            }
+
+            clickEvents.addAll(text.clickEvents);
+            hoverEvents.addAll(text.hoverEvents);
+        }
+
+        return new StyledText(parts, clickEvents, hoverEvents);
+    }
+
+    public static StyledText join(String separator, Iterable<StyledText> texts) {
+        return join(separator, Iterables.toArray(texts, StyledText.class));
+    }
+
+    public static StyledText concat(StyledText... texts) {
+        List<StyledTextPart> parts = new LinkedList<>();
+        List<ClickEvent> clickEvents = new LinkedList<>();
+        List<HoverEvent> hoverEvents = new LinkedList<>();
+
+        for (StyledText text : texts) {
+            parts.addAll(text.parts);
+            clickEvents.addAll(text.clickEvents);
+            hoverEvents.addAll(text.hoverEvents);
+        }
+
+        return new StyledText(parts, clickEvents, hoverEvents);
+    }
+
+    public static StyledText concat(Iterable<StyledText> texts) {
+        return concat(Iterables.toArray(texts, StyledText.class));
+    }
+
+    public StyledText getNormalized() {
+        return new StyledText(
+                parts.stream().map(StyledTextPart::asNormalized).collect(Collectors.toList()),
+                temporaryWorkaround,
+                clickEvents,
+                hoverEvents);
+    }
+
+    public StyledText trim() {
+        return new StyledText(
+                parts.stream().map(StyledTextPart::trim).collect(Collectors.toList()),
+                temporaryWorkaround,
+                clickEvents,
+                hoverEvents);
+    }
+
+    public boolean isEmpty() {
+        return parts.stream().allMatch(StyledTextPart::isEmpty);
```

Comment:
It's empty if there are no parts.

Reply:
If the stream is empty, true is returned.

Reply:
Let's take a step back here. My point was that isEmpty is true iff (if and only if) we have no parts.

But I now realize your code checks if we have an empty part somewhere in our part list. 

I think that should not be allowed. If we can have empty parts in the list, then there is no longer a 1-to-1 relationship with e.g. a coded string and a StyledText, since the styled text can have an arbitrary number of "hidden" empty parts in it.

So we need to make sure this cannot happen. And then this function once again reduced down to checking if we have no parts, since then by definition if we have any parts, they cannot be empty.

Reply:
The check for no empty parts is already in place in suitable locations, when converting from string. I will add the check for components. Are you happy with `isBlank`?

Reply:
Yeah, isBlank is a different story. We can have a sequence of differently colored spaces, and thus we need to check all of them for isBlank.

(In fact, stuff like this and the trim() issue I pointed out would make awesome unit tests)

Reply:
I will continue to add unit tests, but we can never have enough..